src/v/resource_mgmt/memory_sampling.cc

```diff
@@ -0,0 +1,222 @@
+  ss::logger& logger,
+  config::binding<bool> enabled,
+  double first_log_limit_fraction,
+  double second_log_limit_fraction)
+  : _logger(logger)
+  , _enabled(std::move(enabled))
+  , _first_log_limit_fraction(first_log_limit_fraction)
+  , _second_log_limit_fraction(second_log_limit_fraction) {
+    _enabled.watch([this]() { on_enabled_change(); });
+}
+
+void memory_sampling::on_enabled_change() {
+    // We chose a sampling rate of ~3MB. From testing this has a very low
+    // overhead of something like ~1%. We could still get away with something
+    // smaller like 1MB and have acceptable overhead (~3%) but 3MB should be a
+    // safer default for the initial rollout.
+    const size_t sampling_rate = 3000037;
+
+    // Note no logging here as seastar already logs about this
+    if (_enabled()) {
+        if (ss::memory::get_heap_profiling_sample_rate() == sampling_rate) {
+            return;
+        }
+
+        ss::memory::set_heap_profiling_sampling_rate(sampling_rate);
+    } else {
+        if (ss::memory::get_heap_profiling_sample_rate() == 0) {
+            return;
+        }
+
+        ss::memory::set_heap_profiling_sampling_rate(0);
+    }
+}
+
+void memory_sampling::start() {
+    setup_additional_oom_diagnostics();
+
+    // start now if enabled
+    on_enabled_change();
+
+    ssx::spawn_with_gate(_low_watermark_gate, [this]() {
+        return start_low_available_memory_logging();
+    });
+}
+
+ss::future<> memory_sampling::stop() {
+    _low_watermark_cond.broken();
+
+    co_await _low_watermark_gate.close();
+}
+
+memory_sampling::serialized_memory_profile
+memory_sampling::get_sampled_memory_profile() {
+    auto stacks = ss::memory::sampled_memory_profile();
+
+    std::vector<memory_sampling::serialized_memory_profile::allocation_site>
+      allocation_sites;
+    allocation_sites.reserve(stacks.size());
+
+    for (auto& stack : stacks) {
+        allocation_sites.emplace_back(
+          stack.size,
+          stack.count,
+          ssx::sformat("{}", std::move(stack.backtrace)));
+    }
+
+    return memory_sampling::serialized_memory_profile{
+      ss::this_shard_id(), std::move(allocation_sites)};
+}
+
+ss::future<std::vector<memory_sampling::serialized_memory_profile>>
+memory_sampling::get_sampled_memory_profiles(std::optional<size_t> shard_id) {
+    using result_t = memory_sampling::serialized_memory_profile;
+    std::vector<result_t> resp;
+
+    if (shard_id.has_value()) {
+        resp.push_back(co_await container().invoke_on(*shard_id, [](auto&) {
+            return memory_sampling::get_sampled_memory_profile();
+        }));
+    } else {
+        resp = co_await container().map_reduce0(
+          [](auto&) { return memory_sampling::get_sampled_memory_profile(); },
+          std::vector<result_t>{},
+          [](std::vector<result_t> all, result_t result) {
+              all.push_back(std::move(result));
+              return all;
+          });
+    }
+
+    co_return std::move(resp);
```

Comment:
std::move(resp) not needed here as `resp` is _move-eligible_.

I don't think `std::move` is necessarily harmful here but the same pattern with `return` instead of `co_return` is, in that it prevents possible copy-elision (NRVO).

Reply:
Yeah will remove it, just an artifact of me addressing the unneeded make_ready_future.